modules/prompters.py

```diff
@@ -79,46 +79,79 @@ async def _basic(agent: Agent) -> None:
 
 def trim_message_list(messages: List[Message], target_tok_length: int) -> List[Message]:
     """
-    Trim messages by removing each message starting with the 5th message, until
-    the total token length is less than target_tok_length. Include a message in
-    the trimmed portion indicating that the sequence has been trimmed.
+    Trim messages by:
+    1. Always keeping the first 4 messages
+    2. Adding a trim notice message
+    3. Processing remaining messages in batches of 10 from the end
+    4. Including as many complete batches as possible within token limit
 
     Note that this function always uses tiktoken's cl100k base tokenizer, and
     disregards many details about how to use it for message formats.
     TODO: use the correct tokenizer for any given situation, and use it properly.
     """
     enc = tiktoken.get_encoding("cl100k_base")
-    tokens_to_use = target_tok_length - len(
-        enc.encode(notice_retroactively_trimmed_prompt, disallowed_special=())
-    )
-    for msg in messages[:4]:
-        tokens_to_use -= len(enc.encode(msg.content, disallowed_special=()))
-        tokens_to_use -= len(
+
+    def count_message_tokens(msg: Message) -> int:
+        return len(enc.encode(msg.content, disallowed_special=())) + len(
             enc.encode(json.dumps(msg.function_call), disallowed_special=())
         )
 
-    tail_messages_to_use = []
-    for msg in messages[4:][::-1]:
```

Comment:
this previously added messages from most recent to least recent. The new change adds from oldest to newest. If we trim the most recent messages agents will get stuck in loops because when they take actions and append to the messages their prompt doesn't change